src/ldp/AuthenticatedLdpHandler.ts

```diff
@@ -0,0 +1,171 @@
+    return this.requestParser.canHandle(input.request);
+  }
+
+  /**
+   * Handles the incoming request and writes out the response.
+   * This includes the following steps:
+   *  - Parsing the request to an Operation.
+   *  - Extracting credentials from the request.
+   *  - Extracting the required permissions.
+   *  - Validating if this operation is allowed.
+   *  - Executing the operation.
+   *  - Writing out the response.
+   * @param input - The incoming request and response object to write to.
+   *
+   * @returns A promise resolving when the handling is finished.
+   */
+  public async handle (input: { request: HttpRequest; response: HttpResponse }): Promise<void> {
+    let err: Error = null;
+    let operation: Operation = null;
+
+    try {
+      operation = await this.runHandlers(input.request);
+    } catch (error) {
+      err = error;
+    }
+
+    const writeData = { response: input.response, operation, error: err };
+
+    if (!await this.responseWriter.canHandle(writeData)) {
+      throw new UnsupportedError('Writing the generated response is not supported.');
+    }
+
+    return this.responseWriter.handle(writeData);
+  }
+
+  /**
+   * Runs all handlers except writing the output to the response.
+   * This because any errors thrown here have an impact on the response.
+   * @param request - Incoming request.
+   *
+   * @returns A promise resolving to the generated Operation.
+   */
+  private async runHandlers (request: HttpRequest): Promise<Operation> {
+    const op: Operation = await this.runHandler(
+      this.requestParser,
+      request,
+      'Parsing is not supported for the incoming request.'
+    );
+
+    const credentials: Credentials = await this.runHandler(
+      this.credentialsExtractor,
+      request,
+      'Extracting credentials is not supported for the incoming request.'
+    );
+
+    const permissions: PermissionSet = await this.runHandler(
+      this.permissionsExtractor,
+      op,
+      'Identifying permissions is not supported for the given operation.'
+    );
+
+    await this.runHandler(
+      this.authorizer,
+      { credentials, identifier: op.target, permissions },
+      'Permission checking is not supported for the given operation.'
+    );
+
+    await this.runHandler(
+      this.operationHandler,
+      op,
+      'The given operation is not supported.'
+    );
+
+    return op;
+  }
+
+  /**
+   * Checks if a handler can handle a certain input and then lets the handler handle it handily.
+   * If not, a {@link UnsupportedError} gets thrown.
+   * @param handler - Handler to run.
+   * @param data - Data to handle.
+   * @param error - Error message if an error gets thrown.
+   *
+   * @returns The result of the handle function of the handler.
+   */
+  // eslint-disable-next-line class-methods-use-this
+  private async runHandler<T extends AsyncHandler<TIn, TOut>, TIn, TOut>(handler: T, data: TIn, error: string): Promise<TOut> {
```

Comment:
I have no strong opinion; seems like a good idea.

The initial intention was not to really send causes around, more like "I'm a GET handler, this is POST, so no I don't support this" without really explaining why. More like a fall-through mechanism.

Reply:
Yea also made me think of Comunica when writing it. Practically the main issue that might be a concern is the performance of throwing an error (and catching it) vs returning a boolean. Didn't we have performance issues caused by that or am I misremembering?

Reply:
I initially thought we had performance issues because of building up stacktraces, but removing those appeared to not have any significant impact, so no real performance issues after all.

So I would go forward with the rejection-based handler checking.